qiskit/circuit/operation.py

```diff
@@ -0,0 +1,81 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Operation object
+"""
+
+from collections.abc import Sequence
+
+
+class Operation(Sequence):
```

Comment:
The terminology we've used here has varied a bit, but it's probably worth thinking over before we nail this down.

Maybe someone with more compiler experience can weigh in, but I though canonically an `Instruction` was the combination of an `operation` and its `operands` (qargs and cargs), whereas with this PR we'd have an `Operation` as the combination of an `Instruction` and its `operands`. Is there another name we want to use here, or are we set on `Operation`? (We do already have an `InstructionSet` which is ~largely unused but very similar in structure to what's here.)

Reply:
> Thanks for taking this on. Agree this will be much needed as the circuit model evolves going forward.
> 
> > I'm thinking for performance we'll need to update the access patterns before merging this (which increases the scope of the PR quite a bit). The test execution seems to have slowed down a ton (mostly the shor's algorithm tests).
> > ...
> > So I think the next step here is to update all the QuantumCircuit methods to use attribute access and see if that fixes the overhead from **getitem**.
> 
> This definitely increases the scope of this PR. (If the performance impact really is that substantial, would we need to deprecate index-based access or otherwise help users know to migrate their code?)
> 
> Given that the existing access pattern is entirely index-based, would starting with a `NamedTuple` be a reasonable first step? (Alternately, using a tuple internally so that we could still have `def __getitem__(self, idx): return self._tuple[idx]`)

Yeah, I was under the impression that slotted attribute access was the fastest return and that's what I was optimizing for and my previous experience with named tuple was it's named attribute access was slower.  However, as of py 3.8 that no longer seems to be the case (with this pr: https://github.com/python/cpython/pull/10495 ). I did a quick test with this pr applied on python 3.9:

```
from qiskit import QuantumCircuit
from collections import namedtuple

qc = QuantumCircuit(2)
qc.h(0)

Op = namedtuple('Op', ['instruction', 'qargs', 'cargs'])

op = qc.data[1]
op_tuple = tuple(op)
op_named_tuple = Op(*op_tuple)
```

With indexed access:

```
%timeit op[0]
75.5 ns ± 0.201 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

```
%timeit op_tuple[0]
22.6 ns ± 0.0474 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

```
%timeit op_named_tuple[0]
23 ns ± 0.0631 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

with attribute access:
```
%timeit op.instruction
28.1 ns ± 0.11 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

```
%timeit op_named_tuple.instruction
26.8 ns ± 0.298 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

So I'm thinking the best choice is a named tuple despite my earlier hesitation (which was clearly wrong). I'll have to see how things behave in py3.7 and 3.6 too (although this is the last release with 3.6 support) but I doubt it will be an issue and even if it causes a regression I think it'll be an ok tradeoff (or we can wait until after 0.19 to do this and then we'll only support 3.7, 3.8, 3.9, and 3.10).

Reply:
It just seemed like the logical choice to me since we're already using instruction to represent the opcode. The terminology here is a bit fuzzy, like I've heard instructions and operation (well really opcode) used interchangeable. If you think operation here would be too overloaded the other term I was thinking of was `Statement` (ie one line in assembly). But I'll defer to what others like I don't feel too strongly as long as we're not reusing a name already used (since deprecating `Instruction` as it's used today isn't really a viable path).

Reply:
Yeah I think `Instruction` is the most appropriate name to refer to the tuple, which is like one line of qasm. It's unfortunate it wasn't used like that since the beginning. But there are places in qiskit where it's used like that, e.g. `InstructionScheduleMap`.

I was thinking in this release we can deprecate `Instruction` as it is now, and call it something else (actually Operation is a good name for that: https://en.wikipedia.org/wiki/Quantum_operation). And then re-introduce `Instruction` in the next release to refer to the tuple (Operation, qubits, clbits). This would abide by our 3-month deprecation policy I think.

Reply:
I also prefer `Instruction` for this class, and `Operation` for the base class currently-known-as-Instruction.